pkg/state/scale/consul/consul.go

```diff
@@ -0,0 +1,167 @@
+	eventsPath       string
+	latestEventsPath string
+	gcThreshold      int64
+	logger           zerolog.Logger
+
+	kv *api.KV
+}
+
+func NewStateBackend(log zerolog.Logger, path string) scale.Backend {
+	consul, _ := client.NewConsulClient()
+
+	return &StateBackend{
+		basePath:         path + "state/",
+		eventsPath:       path + "state/" + "events/",
+		latestEventsPath: path + "state/" + "latest-events/",
+		gcThreshold:      scale.GarbageCollectionThreshold,
+		logger:           log,
+		kv:               consul.KV(),
+	}
+}
+
+func (s StateBackend) GetScalingEvents() (map[uuid.UUID]map[string]*state.ScalingEvent, error) {
+	kv, _, err := s.kv.List(s.eventsPath, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if kv == nil {
+		return nil, nil
+	}
+
+	out := make(map[uuid.UUID]map[string]*state.ScalingEvent)
+
+	for i := range kv {
+		keyState := &state.ScalingEvent{}
+
+		if err := json.Unmarshal(kv[i].Value, keyState); err != nil {
+			return nil, errors.Wrap(err, "failed to unmarshal Consul KV value")
+		}
+
+		keySplit := strings.Split(kv[i].Key, "/")
+
+		id, err := uuid.FromString(keySplit[len(keySplit)-2])
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get UUID from string")
+		}
+
+		out[id] = map[string]*state.ScalingEvent{keySplit[len(keySplit)-1]: keyState}
+	}
+
+	return out, nil
+}
+
+func (s StateBackend) GetScalingEvent(id uuid.UUID) (map[string]*state.ScalingEvent, error) {
+	kv, _, err := s.kv.List(s.eventsPath+id.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if kv == nil {
+		return nil, nil
+	}
+
+	out := make(map[string]*state.ScalingEvent)
+
+	for i := range kv {
+		s := &state.ScalingEvent{}
+		if err := json.Unmarshal(kv[i].Value, s); err != nil {
+			return nil, errors.Wrap(err, "failed to unmarshal Consul KV value")
+		}
+
+		keySplit := strings.Split(kv[i].Key, "/")
+		out[keySplit[len(keySplit)-1]] = s
+	}
+
+	return out, nil
+}
+
+func (s StateBackend) PutScalingEvent(job string, event *state.ScalingEventMessage) error {
+
+	sEntry := &state.ScalingEvent{
+		EvalID:  event.EvalID,
+		Source:  event.Source,
+		Time:    event.Time,
+		Status:  event.Status,
+		Details: state.EventDetails{Count: event.Count, Direction: event.Direction},
+	}
+
+	marshal, err := json.Marshal(sEntry)
+	if err != nil {
+		return err
+	}
+
+	// Write the event to the general store.
+	ePair := &api.KVPair{
+		Key:   fmt.Sprintf("%s%s/%s:%s", s.eventsPath, event.ID.String(), job, event.GroupName),
+		Value: marshal,
+	}
+	if _, err = s.kv.Put(ePair, nil); err != nil {
+		return err
+	}
+
+	// Write the new event to the latest store.
+	lePair := &api.KVPair{
+		Key:   fmt.Sprintf("%s%s/%s:%s", s.latestEventsPath, event.ID.String(), job, event.GroupName),
+		Value: marshal,
+	}
+	_, err = s.kv.Put(lePair, nil)
+	return err
+}
+
+func (s StateBackend) RunGarbageCollection() {
+	kv, _, err := s.kv.List(s.eventsPath, nil)
+	if err != nil {
+		s.logger.Error().Err(err).Msg("GC failed to list events in backend store")
```

Comment:
Does this not return because kv would be nil if theres an error?

Reply:
Yes, there is also one scenario where an error wont be returned and kv will be nil which is caught on line 139.